include/ros_alarms/heartbeat_monitor.hpp

```diff
@@ -0,0 +1,219 @@
+
+  // Register cleared alarm w/ alarm server
+  __alarm_broadcaster.clear();
+
+  // For better error msgs
+  std::stringstream obj;
+  obj << "HeartbeatMonitor[heartbeat_name=" << heartbeat_topic
+      << ", alarm_name=" << __alarm_proxy.alarm_name
+      << ", node=" << __alarm_proxy.node_name << "]";
+  __object_name = obj.str();
+
+  std::stringstream init_msg;
+  init_msg << "Node " << __alarm_proxy.node_name << " is now monitoring the following"
+    " heartbeat: " << __heartbeat_topic;
+  //ROS_INFO("%s", init_msg.str().c_str());
+}
+
+template <typename msg_t>
+void HeartbeatMonitor<msg_t>
+::__heardHeartbeat(msg_t beat_msg)
+{
+  auto receipt_time = ros::Time::now();
+  auto last_beat_time  = __last_beat;
+  bool valid_beat = false;
+
+  // Call predicate and set last healthy beat time
+  try
+  {
+     valid_beat = __msg_predicate(beat_msg);
+  }
+  catch(std::exception &e)
+  {
+     std::stringstream err;
+     err << __object_name << " - Predicate function threw an exception: " << e.what();
+     ROS_WARN("%s", err.str().c_str());
+  }
+
+  if(!valid_beat) // Don't do stuff if beat is invalid
+    return;
+
+  __last_beat = receipt_time; // update __last_beat
+  auto time_since_prev_beat = receipt_time - last_beat_time;
+
+  // Set the last healthy beat time as a json param on the alarm proxy
+  std::stringstream json;
+  // not exactly sure if this will work with json.loads()
+  json << "{last_heathy_beat : " <<  __last_beat.toSec() << "}";
+  __alarm_proxy.json_parameters = json.str();
+
+  if(__healthy)  // We're done if the heartbeat is healthy
+    return;
+
+  // Now we will determine if we need to start recovering or increase the
+  //   recovery counter or if we are fully recovered and need to clear alarm
+
+  if(__recovering)  // Either increase recovery time or reset it
+  {
+    if(time_since_prev_beat < __time_to_raise)
+    {
+      __time_recovering += time_since_prev_beat;
+      if(__time_recovering > __time_to_raise)
+      {
+        // Reset health flag and exit recovery mode
+        __healthy = true;
+        __recovering = false;
+
+        // Clear heartbeat-loss alarm
+        __alarm_broadcaster.clear();
+      }
+    }
+    else
+    {
+      __time_recovering = ros::Duration(0);
+    }
+  }
+  else  // Set flag for recovery mode
+  {
+    __recovering = true;
+    __time_recovering = ros::Duration(0.0);
+  }
+}
+
+template <typename msg_t>
+void HeartbeatMonitor<msg_t>
+::__checkForHeartbeatLoss(const ros::TimerEvent &te)
+{
+  if(!__healthy) // Can't lose what's already lost
+    return;
+
+  ros::Duration time_since_last_beat = ros::Time::now() - __last_beat;
+
+  if(time_since_last_beat > __time_to_raise) // Raise heartbeat-loss alarm
+  {
+    __alarm_proxy.raised = true;
+    __alarm_broadcaster.raise();
+    __healthy = false;
+  }
+  // Should we add a mode where we increase the severity based on the time the
+  //   heartbeat has been lost?
```

Comment:
That does sound like a good idea, if you want to add it to the c++ implementation, I can add it to the python at some point soon. I think a heartbeat alarm is a good way to implement that, with a predicate that checks for jumps/covariance increases that can trigger alarms as well.